PostSharp.Community.StructuralEquality.Weaver/EqualsInjection.cs

```diff
@@ -0,0 +1,408 @@
+        private readonly IGenericMethodDefinition instanceEqualsMethod;
+        private readonly IGenericMethodDefinition staticEqualsMethod;
+        private readonly IGenericMethodDefinition collectionEqualsMethod;
+        private readonly IGenericMethodDefinition getTypeFromHandleMethod;
+        private readonly ITypeSignature equatableInterface;
+
+        public EqualsInjection( Project project )
+        {
+            this.objectType = project.Module.Cache.GetIntrinsic( IntrinsicType.Object );
+            this.booleanType = project.Module.Cache.GetIntrinsic( IntrinsicType.Boolean );
+
+            this.objectTypeDef = this.objectType.GetTypeDefinition();
+
+            this.referenceEqualsMethod = project.Module.FindMethod( objectTypeDef, "ReferenceEquals" );
+            this.instanceEqualsMethod = project.Module.FindMethod( objectTypeDef, "Equals", 1 );
+            this.staticEqualsMethod = project.Module.FindMethod( objectTypeDef, "Equals", 2 );
+            this.getTypeMethod = project.Module.FindMethod( objectTypeDef, "GetType" );
+
+            var collectionHelperTypeDef = project.Module.Cache.GetType( typeof(CollectionHelper) ).GetTypeDefinition();
+            this.collectionEqualsMethod = project.Module.FindMethod( collectionHelperTypeDef, "Equals",
+                declaration => declaration.IsStatic );
+
+            var typeTypeDef = project.Module.Cache.GetType( typeof(Type) ).GetTypeDefinition();
+            this.getTypeFromHandleMethod = project.Module.FindMethod( typeTypeDef, "GetTypeFromHandle" );
+
+            this.equatableInterface = project.Module.Cache.GetType( typeof(IEquatable<>) );
+        }
+
+        public void AddEqualsTo( TypeDefDeclaration enhancedType, StructuralEqualityAttribute config,
+            ISet<FieldDefDeclaration> ignoredFields )
+        {
+            var typedEqualsMethod = this.InjectEqualsType( enhancedType, config, ignoredFields );
+            this.InjectEqualsObject( enhancedType, config, typedEqualsMethod );
+
+            this.AddEquatable( enhancedType );
+        }
+
+        private void AddEquatable( TypeDefDeclaration enhancedType )
+        {
+            var genericInstance = this.equatableInterface.GetTypeDefinition().GetGenericInstance( new GenericMap(
+                enhancedType.Module,
+                new[] {enhancedType.GetCanonicalGenericInstance()} ) );
+            enhancedType.InterfaceImplementations.Add( genericInstance.TranslateType( enhancedType.Module ) );
+        }
+
+        private MethodDefDeclaration InjectEqualsType( TypeDefDeclaration enhancedType,
+            StructuralEqualityAttribute config, ICollection<FieldDefDeclaration> ignoredFields )
+        {
+            // public virtual bool Equals( Typed other )
+            var equalsDeclaration = new MethodDefDeclaration
+            {
+                Name = "Equals",
+                Attributes = MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.Virtual,

```

Comment:
You chose to have the strongly-typed method virtual and to call its base method. I was considering instead calling the base Equals(object) method.

I see the benefit of virtualizing the strongly-typed method is that you bypass the possibly strong type check in the base class which might prevent the overridding Equals from working, and also that you gain more performance. I see the downside as the fact that the base class must implement IEquatable (otherwise why would it have a strongly-typed Equals) for the base.Equals call to work.

I don't know what's better.